Add new PortAllocator

dperny opened the pull request:
This is the first commit in a series of commits rewriting the network allocator (#2516).

This is a rewrite of the PortAllocator component. The PortAllocator is a very simple component that keeps track of which ingress network ports are in use in the cluster.

This new PortAllocator is not yet wired up, because it exposes a completely different interface compared to the existing one. It will be wired up as the rewrite progresses. It has been put here for a review, and can be reviewed as if it is to be merged, but it *should not* be merged at this time.

Because the PortAllocator is a simple, serial object, with a well-defined interface, I've chosen to use Ginkgo to write tests, as it has helped me enumerate the test cases I need to write and focus on testing those cases. I'm not suggesting we use Ginkgo anywhere else, and I am especially not suggesting we port any existing tests to Ginkgo. It just happens to be the right tool for this job.

At this time, the main design of PortAllocator is finished, and it only needs some more tests written.

Includes a vendoring of Ginkgo, which makes this PR look bigger than it is.

stevvooe commented on manager/allocator/network/port/port.go:
```diff
@@ -0,0 +1,367 @@
+const (
+	// DynamicPortStart is the start of the dynamic port range from which node
+	// ports will be allocated when the user did not specify a port.
+	DynamicPortStart uint32 = 30000
```
I think you might be able to get away with an `int16`, although I wouldn't recommend using an unsigned here.

stevvooe replied:
Rather, you can just use a regular `int`.

dperny replied:
I use `uint32` because that's the type of `PortConfig.PublishPort`.

stevvooe commented on manager/allocator/network/port/port.go:
```diff
@@ -0,0 +1,367 @@
+}
+
+// Commit commits a change to the PortAllocator. There are no safety checks.
+func (pa *PortAllocator) Commit(p *Proposal) {
```
I would recommend moving the `Commit` method to the `Proposal` object itself. Checkout the bolt db api for an example of how this would work.

Can this method have an error?

dperny replied:
I would like to avoid this method having an error. I'd like to avoid having this method. It's literally just because there's no easy way to roll back to a previous PortAllocator state if you abandon allocation.

stevvooe commented on manager/allocator/network/port/port.go:
```diff
@@ -0,0 +1,367 @@
+// as a map key
+type port struct {
+	// protocol represented by this port space
+	protocol api.PortConfig_Protocol
```
You can save a little space by having a struct with a port vector for each protocol:

```
type PortAllocator struct {
  udp map[port]struct{}
  tcp map[port]struct{}
  sctp map[port]struct{}
}
```

stevvooe replied:
It's possible map overhead removes this benefit, however.

Like we discussed, this is _probably_ sparse. I did some experimentation with using `big.Int` as a bit set and that is pretty minimal.

dperny replied:
It's easier to reason in the code about ports-as-a-value than it is to reason different-protocols-different-spaces. 

The overhead of the extra data in each port object is minimal compared to the overhead of literally all the other dumb inefficient crap I'm doing, so even if we optimized here, it would be the tiniest win.

dperny commented:
Updated with more tests, changed the Proposal to have a Commit method.

dperny commented:
I realized that mutating the endpoint object in `AllocateEndpoint` was dumb as hell, so I changed the interface for that method to return (as part of the Proposal) a slice of PortConfigs, which the caller should stick into the endpoint. Less mutation is a good thing.

dperny commented:
I don't think adding Ginkgo as a separate PR is helpful. I could agree that putting the vendoring into a separate commit might be though.

dperny commented:
It's always a good feeling when you're writing tests to catch an error in your code.

dperny commented:
91.8% test coverage of the new code. I just need to figure out what currently isn't tested.

dperny commented:
100.0% test coverage. My work here is done.

dperny commented:
I wrote a utility to spit out the specs of what's being tested. Here's the summary:
```
port.Allocator
    adding a new endpoint
        to a clean port allocator
            - should succeed
            - should not modify the spec
            - should not modify the endpoint
            - should not modify the Allocator state before calling Commit
            - should be idempotent
        to an allocator with ports in use
            with colliding ports in the new endpoint
                - should return ErrPortInUse
            with no colliding ports in the new endpoint
                - should succeed
                - should return ports matching the spec
            with colliding port numbers, but different protocols
                - should succeed
                - should return ports matching the spec
        with invalid publish or target ports
            - should return ErrInvalidEndpoint
    updating an endpoint
        to add ports
            if ports collide
                - should fail with ErrPortInUse
            if ports do not collide
                - should succeed
                - should add return ports matching the spec's ports
        to remove ports
            - should succeed
            - should return ports matching the spec's ports
            - should free the removed ports for another object to use
            - should not free any ports not removed
        to change name and target port
            - should be a noop
            - should return ports matching the spec's ports
        to change published port
            - should allocate the new port
            - should free the old port
            - should not free any other ports
        to change protocol
            - should allocate the new port
            - should free the old port
    dyanmic port allocation
        adding a port with no PublishPort specified
            - should succeed
            - should assign a port from the dynamic port range
            - should be idempotent
            - should not modify the spec
            - should cause new allocations to take the next available port
        when the dynamic port space is exhausted
            - should fill all the way up successfully
            - should not prevent dynamic allocation in a different port space
            adding a port to the same space
                - should return ErrPortSpaceExhausted
        when a port in the dynamic range desired to be statically allocated
            - should succeed
            - should give the statically allocated port first priority
        when changing ports
            from dynamic to statically allocated
                - should succeed
            from statically to dynamically allocated
                - should succeed
                - should free the old static allocation
                - should dynamically allocate a new port number
    endpoints with publish mode host ports
        when adding a host port
            - should succeed
            - should be a noop
            - should include the host ports in the returned ports
            - should not mark any ports in use
            - should be idemopotent
            when adding another host port with no PublishPort specified
                - should succeed
                - should be a noop
                - should not fill in the published port
    deallocate
        - should return a proposal that isn't a noop but has no ports
        - should not modify the spec
        - should free any ports in the endpoint
        - should only free endpoints of the correct protocol
        - should not free any ports with publish mode host
```

dperny commented:
rebased and added `doc.go`.

cyli commented on manager/allocator/network/port/port.go:
```diff
@@ -0,0 +1,429 @@
+		// one available
+		for i := DynamicPortStart; i <= DynamicPortEnd; i++ {
+			portObj.port = i
+			if _, ok := pa.ports[portObj]; !ok {
```
This doesn't seem to take into account any deallocated ports.  If we're near almost exhausting the port space, and this update deallocates 3 ports but needs to allocate 3 different ports, we should be able to do so (at least, by share number) but we can't,  since this loop isn't checking to see if the taken port has been deallocated.

I can also see how maybe that makes sense to do because if we're changing the name of the port or something else about the port config, etc., maybe it'd be confusing if the changed port config gets dynamically allocated the same port on update.

dperny replied:
Actually, yeah, you're right. I'll need to figure out the correct solution.

dperny replied:
Fixed this issue by adding another loop further down checking if any dynamic ports are freed.

cyli replied:
Totally non-blocking nit - that loop looks fine and definitely fixes the issue - I was just wondering if it'd be easier to understand as an extra else block here?  Since this loop already iterates over only the dynamic ports, that would mean that you may not need an extra check for whether the deallocated port is in the dynamic ports?

Also you already have a port object you can use to index into the `deallocate` map.

This is probably just a disfunction of how my brain works, but I find `goto` (or `continue <label>`) control flow a little more confusing to follow than just if else blocks.

dperny replied:
doing this in two loops preserves the behavior of preferring to not reassign the same dynamic ports that were just freed over again. i don't think that that behavior is _required_, and i don't want to make that behavior _part of the api contract_, but it is rather nice. i can add a comment explaining that we try this but not to rely on it.

> Also you already have a port object you can use to index into the deallocate map.

I'm brain farting on what you mean by this.

You're right about `continue`, though, it is a slightly-less-harmful goto.

cyli replied:
> doing this in two loops preserves the behavior of preferring to not reassign the same dynamic ports that were just freed over again. i don't think that that behavior is required, and i don't want to make that behavior part of the api contract, but it is rather nice. i can add a comment explaining that we try this but not to rely on it.

Ah that makes sense.

> I'm brain farting on what you mean by this.

Er sorry, I meant if you had the one loop, you could just index into `deallocated` with https://github.com/docker/swarmkit/pull/2579/files#diff-50dd126d07c8b730e5bae674f9542aa2R384.

Not really relevant if we want a separate loop as you say to try to change the dynamically allocated ports each time.

cyli commented on manager/allocator/network/port/port_test.go:
```diff
@@ -0,0 +1,955 @@
+				endpointCopy := endpoint.Copy()
+				endpointCopy.Spec = spec
+				endpointCopy.Ports = p.Ports()
+				p.Commit()
```
Non-blocking:  would it make sense to add an extra check here before the commit to assert it's not a noop?

dperny replied:
Sure, this test can actually probably be reworked a bit to verify that allocator state changes AFTER commit as well, but I thought I had that test someplace else...

dperny replied:
Added a check for not noop to the "should succeed" case.

cyli commented on manager/allocator/network/port/port_test.go:
```diff
@@ -0,0 +1,955 @@
+		})
+		// Pending, we need to test this but I'm not sure what the appropriate
+		// behavior should be
+		PContext("to change publish mode", func() {
```
Are we still deciding what the appropriate behavior should be?  It seems like nothing gets deallocated or allocated (so if it previously had an allocated port, changing the publish mode to host would not deallocate the previous port).

dperny replied:
Must've missed this when I was going through, I will add these test cases, thanks!

cyli commented on manager/allocator/network/port/port_test.go:
```diff
@@ -0,0 +1,955 @@
+					Expect(e).ToNot(HaveOccurred())
+					Expect(p).ToNot(BeNil())
+					Expect(p.IsNoop()).ToNot(BeTrue())
+					Expect(p.Ports()).ToNot(BeEmpty())
```
Should we add more assertions that the previous dynamic allocation was freed?

dperny replied:
Added this assertion.

dperny commented on manager/allocator/network/port/doc.go:
```diff
@@ -0,0 +1,28 @@
+// no PublishedPort.
+//
+// Notably, the allocator only keeps track of which ports are in use, not which
+// servies they are in use by. For the port allocator to function correctly,
```
good catch.

dperny commented on manager/allocator/network/port/port.go:
```diff
@@ -0,0 +1,429 @@
+	Commit()
+
+	// IsNoop returns true if the Proposal doesn't actually modify allocator
+	// state This does not mean the endpoint object has not been modified, just
```
Another good catch, will clarify.

dperny commented on manager/allocator/network/port/port.go:
```diff
@@ -0,0 +1,429 @@
+	return prop.ports
+}
+
+// Commit commits the proposal to the port allocator. Part of our
```
I have no idea what you

dperny commented on manager/allocator/network/port/port.go:
```diff
@@ -0,0 +1,429 @@
+// Commit commits the proposal to the port allocator. Part of our
+func (prop *proposal) Commit() {
+	if prop.IsNoop() {
+		// cannot commit with nil port allocator
```
You're right. I'll remove the check. I probably had the check in IsNoop and removed it.

dperny commented on manager/allocator/network/port/port.go:
```diff
@@ -0,0 +1,429 @@
+}
+
+func (prop *proposal) IsNoop() bool {
+	// if every port in allocate is in deallocate, and every port in deallocate
```
We can't check the length, because the actual ports in use may have changed. The comment can be reworded though.

cyli replied:
Oops sorry I meant do a length check, then a single loop as opposed to 2 loops.  If they are the same length and every port in allocate is also in reallocate they must have the same keys I think

dperny commented on manager/allocator/network/port/doc.go:
```diff
@@ -0,0 +1,28 @@
+// network ports. Its purpose is to assign ports to Endpoint objects based on
+// the provided Spec. For ingress ports, it keeps track of the availability of
+// ports, and handles dynamic allocation of ports when the user has specified
+// no PublishedPort.
```
Yes, we do.

dperny commented:
Thanks for the excellent review!

> If I had a previously allocated port, but changed the publish mode to host? Is the previously allocated port never deallocated?

It should be, actually, but there's no test case for it, so I'll add one and make sure.

> If we're near exhausting the port space, and I update the endpoint to change the name of the dynamically allocated ports, will it allocate me the new ports or will I get an exhaustion error?

You're right, actually, that it will give an exhaustion error, I think. I'm not sure, so I'll write a test case.

dperny commented:
Updated PR:
* Added tests for changing port publish mode
* Fixed issue where trying to change ports when the dynamic port range is exhausted would fail
* Added more assertions
* Reworded some comments

cyli commented on manager/allocator/network/port/port.go:
```diff
@@ -0,0 +1,451 @@
+	// object, which contains the proposed changes to the Allocator. this
+	// means if there is a failure in the caller after calling Allocate, the
+	// caller can discard the changes
+	prop := &proposal{
```
Non-blocking:  I was wondering if we could just call Deallocate here, and modify the proposal that it returned with an `allocate` map?  Since otherwise the code is the same.

dperny commented:
split into two commits, one vendoring Ginkgo and the other adding the port allocator.

Adirio commented on manager/allocator/network/port/port.go:
```diff
@@ -0,0 +1,451 @@
+	//
+	// Used mostly for testing, and should probably not be relied on outside of
+	// this package.
+	IsNoop() bool
```
If you think it should not be trusted from outside, maybe not exporting it would be wise.

dperny replied:
I'm going to change this comment actually. It is probably a useless function outside of the tests, but it should be able to be relied on just fine.

Adirio commented on manager/allocator/network/port/port.go:
```diff
@@ -0,0 +1,451 @@
+
+// Deallocate takes an endpoint and provides a Proposal that will deallocate
+// all of its ports.
+func (pa *Allocator) Deallocate(endpoint *api.Endpoint) Proposal {
```
Should this function admit variadic arguments? `func (pa *Allocator) Deallocate(endpoints ...*api.Endpoint) Proposal {` Would only need using a `for` loop wrapping the current one and will allow current behaviour but would optimize deallocating multiple endpoints by grouping them in a single `Proposal`. May you ever need multiple endpoints?

dperny replied:
Nah, because you very rarely are deallocating many endpoints at once. You're usually deallocating them one-by-one. To make use of a variadic, you'd have to write the caller to do batching. It's not worthwhile.

Adirio replied:
Understood

Adirio commented on manager/allocator/network/port/port.go:
```diff
@@ -0,0 +1,451 @@
+		}
+		if spec.TargetPort > masterPortEnd {
+			return nil, ErrInvalidEndpoint{fmt.Sprintf("target port %v isn't in the valid port range", spec.TargetPort)}
+		}
```
These are the checks that could be avoided if the underlying type was `uint16`.

Adirio commented on manager/allocator/network/port/port.go:
```diff
@@ -0,0 +1,451 @@
+
+type proposal struct {
+	pa         *Allocator
+	ports      []*api.PortConfig
```
Is `ports` and the getter function `*proposal.Ports()` being used for anything? It seems they are not needed, maybe some deprecated code?

dperny replied:
Yes. it's the most important part, used by the caller. The alternative to `Ports()` would be to return `(Proposal, []*api.PortConfig, error)` from `Allocate`. That might be cleaner. WDYT?

Adirio replied:
Oh, I see, you need to modify the given []*api.PortConfig (e.g. dynamic assigned ports) but you can not do it over the given pointer as the whole two-phase model would get broken.

Between the two proposed options, I would suggest sticking to the first because all the information is contained in the proposal, you do not have to keep two variables to keep information that does not have any meaning by its own.

Adirio commented on manager/allocator/network/port/port.go:
```diff
@@ -0,0 +1,451 @@
+
+// Restore adds the current endpoints to the local state of the port allocator
+// but does not perform any new allocation.
+func (pa *Allocator) Restore(endpoints []*api.Endpoint) {
```
Am I correct thinking that this method is only used just after creating the allocator? Taking it to a normal function instead of a "method", creating the allocator inside and returning the allocator looks nicer in my head, as it will ensure that it only gets called once on an empty allocator.

dperny replied:
there are a few good reasons not to do it like this:

1. the constructor never returns an error, which means that anything that relies on it can in turn avoid returning errors from its constructor, all the way to the top. this is handy because when everything is said and done, it means you can have the top-level allocator component initialize the subcomponents in its constructor, and then run the Restore method in its Run function. it's hard to explain

2. you can call Restore multiple times, which isn't always necessary, but which is useful for testing because subtests can call Restore again with more state.

Adirio commented on manager/allocator/network/port/port.go:
```diff
@@ -0,0 +1,451 @@
+	// Used mostly for testing, and should probably not be relied on outside of
+	// this package.
+	IsNoop() bool
+}
```
You are creating an interface in order not to export the type itself so that `proposals` can not be created directly. I would go even further and create an `Allocator` interface and extract both interfaces to the `allocator` package so that all allocators need to follow these interfaces. That would requirte some generalization.

Something in the lines of:
```
type Allocator interface {
    Allocate(old, new *api.Task) (Proposal, error)
    Deallocate(*api.Task) (Proposal, error)
}

type Proposal interface {
    Commit()
    IsNOOP() bool
}
```
And then if you want to still hide the proposal object you can make a specialized interface inside the `ports` subpackage:
```
type PortProposal interface {
    allocator.Proposal
    Ports() []*api.PortConfig
}
```

Adirio commented on manager/allocator/network/port/port.go:
```diff
@@ -0,0 +1,451 @@
+					portObj.port = deallocated.port
+					p.PublishedPort = deallocated.port
+					prop.allocate[portObj] = struct{}{}
+					continue ports
```
What about removing it from `deallocate` instead of inserting it into `allocate`?
This would remove the need to check if it is in the allocate map, reduce the iterations in this loop to get to a solution and also simplify the commit by removing one add and one remove task.

dperny commented on manager/allocator/network/port/port.go:
```diff
@@ -0,0 +1,451 @@
+	// protocol represented by this port space
+	protocol api.PortConfig_Protocol
+	port     uint32
+}
```
no, i'm not. the allocator code is entirely contained within itself. i don't think there's anything to be gained by deduplicating this small struct across those two components, though.

dperny commented:
Ah, shit, I pushed to the wrong branch. My naming system is terrible. Anyway, closing this in favor of a pending PR.
